annotation_pipeline/segment.py

```diff
@@ -1,101 +1,149 @@
-            mzs_list.append(mzs_)
-            ints_list.append(ints_)
-            estimated_size_mb += 2 * mzs_.nbytes + ints_.nbytes
-            curr_sp_i += 1
-            if estimated_size_mb > max_size:
-                yield sp_inds_list, mzs_list, ints_list
-                sp_inds_list, mzs_list, ints_list = [], [], []
+        # Iterate in the same order that intensities are laid out in the file, hopefully this will
+        # prevent fragmented read patterns
+        for sp_i in np.argsort(imzml_reader.intensityOffsets):
+            spectrum_size = imzml_reader.mzLengths[sp_i] * row_size
+            if estimated_size_mb + spectrum_size > MAX_CHUNK_SIZE:
                 estimated_size_mb = 0
+                yield np.array(chunk_sp_inds)
+                chunk_sp_inds = []
+
+            estimated_size_mb += spectrum_size
+            chunk_sp_inds.append(sp_i)
+
+        if chunk_sp_inds:
+            yield np.array(chunk_sp_inds)
+
+    def upload_chunk(ibm_cos, ch_i):
+        chunk_sp_inds = chunks[ch_i]
+        # Get imzml_reader from COS because it's too big to include via pywren captured vars
+        imzml_reader = pickle.loads(read_object_with_retry(ibm_cos, config["storage"]["ds_bucket"],
+                                              input_data["ds_imzml_reader"]))
+        n_spectra = sum(imzml_reader.mzLengths[sp_i] for sp_i in chunk_sp_inds)
+        sp_mz_int_buf = np.zeros((n_spectra, 3), dtype=imzml_reader.mzPrecision)
+
+        chunk_start = 0
+        for sp_i, mzs, ints in get_spectra(input_data['ibd_path'], imzml_reader, chunk_sp_inds):
+            chunk_end = chunk_start + len(mzs)
+            sp_mz_int_buf[chunk_start:chunk_end, 0] = sp_id_to_idx[sp_i]
+            sp_mz_int_buf[chunk_start:chunk_end, 1] = mzs
+            sp_mz_int_buf[chunk_start:chunk_end, 2] = ints
+            chunk_start = chunk_end
+
+        by_mz = np.argsort(sp_mz_int_buf[:, 1])
+        sp_mz_int_buf = sp_mz_int_buf[by_mz]
+        del by_mz
 
-        if len(sp_inds_list) > 0:
-            yield sp_inds_list, mzs_list, ints_list
-
-    def _upload_chunk(ch_i, sp_mz_int_buf):
         chunk = msgpack.dumps(sp_mz_int_buf)
         key = f'{input_data["ds_chunks"]}/{ch_i}.msgpack'
         size = sys.getsizeof(chunk) * (1 / 1024 ** 2)
         logger.info(f'Uploading spectra chunk {ch_i} - %.2f MB' % size)
-        cos_client.put_object(Bucket=config["storage"]["ds_bucket"],
-                              Key=key,
-                              Body=chunk)
+        ibm_cos.put_object(Bucket=config["storage"]["ds_bucket"],
+                           Key=key,
+                           Body=chunk)
         logger.info(f'Spectra chunk {ch_i} finished')
         return key
 
-    max_size = 512 * 1024 ** 2  # 512MB
-    chunk_it = chunk_size(coordinates, max_size)
+    chunks = list(plan_chunks())
+    memory_capacity_mb = 3072
+    futures = pw.map(upload_chunk, range(len(chunks)), runtime_memory=memory_capacity_mb)
```

Comment:
This was weird - `chunks` was ~5MB with the bigger datasets, and PyWren raised an error saying the data size exceeded 4MB if I passed the chunks as arguments (i.e. `pw.map(upload_chunk, enumerate(chunks))`). However, it let me transfer `chunks` to the actions as a captured variable. Now all actions have to download the full `chunks` list and pick their appropriate item...

I understand why PyWren rejects large tasks, but in this case it has actually pushed me toward slightly less efficient code.

Reply:
in this case, PyWren uploads each chunk into COS and then each action downloads its relevant chunk again, maybe a better approach will be to store these values inside COS (you can even use cloudobjects for convenience) from the beginning when you generate them in `plan_chunks()`

Reply:
These values don't get much larger, so it's probably not worth it.